Bluetooth: hci: added h4_close function to close bluetooth interface

lfilliot opened the pull request:
Inside the hci driver API in h4.c, there were no API function allowing to close the bluetooth interface, as there were close function defined for other bluetooth hci driver. The function disables Bluetooth reception and sending, and calls a weak function allowing to implement user specific behaviors while shutting down bluetooth interface

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -521,7 +521,7 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
 	k_tid_t tid;
 
-	LOG_DBG("");
+	LOG_DBG("h4_open");
```
the function name should already be generated with `""`

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,33 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
 	return 0;
 }
 
+int __weak bt_hci_transport_setup_down(const struct device *dev)
```
```suggestion
int __weak bt_hci_transport_teardown(const struct device *dev)
```

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,33 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+
+	int ret;
+
+	LOG_DBG("h4_close");
```
```suggestion
	LOG_DBG("");
```

lfilliot replied:
Fixed (commit amended)

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,33 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+{
+	const struct h4_config *cfg = dev->config;
+
+	int ret;
```
```suggestion
	int err;
```

lfilliot replied:
Fixed (commit amended)

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,33 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+	uart_irq_rx_disable(cfg->uart);
+	uart_irq_tx_disable(cfg->uart);
+
+	ret = bt_hci_transport_setup_down(cfg->uart);
+	if (ret < 0) {
+		return -EIO;
+	}
```
```suggestion
	err = bt_hci_transport_teardown(cfg->uart);
	if (err) {
		return err;
	}
```

carlescufi commented:
Thanks for the PR! Please take a look at [our commit message guidelines](https://docs.zephyrproject.org/latest/contribute/guidelines.html#commit-message-guidelines) to find out how to format your commit messages, and at [our contribution workflow](https://docs.zephyrproject.org/latest/contribute/guidelines.html#contribution-workflow) to understand how to update your Pull Request.

cvinayak commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,33 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+static int h4_close(const struct device *dev)
+{
+	const struct h4_config *cfg = dev->config;
+
```
```suggestion
```

lfilliot replied:
Done

cvinayak reviewed (commented):
One more minor nit-pick, but otherwise LGTM.

lfilliot committed:
Bluetooth: hci: added h4_close function

Inside the hci driver API in h4.c, there were no
API function allowing to close the bluetooth
interface, as there were close function defined
for other bluetooth hci driver.

The function disables Bluetooth reception and
sending, and calls a weak function allowing
to implement user specific behaviors
while shutting down bluetooth interface

Signed-off-by: FILLIOT Louis <l.filliot@lacroix.group>

lfilliot commented on drivers/bluetooth/hci/h4.c:
```diff
@@ -545,6 +545,32 @@ static int h4_open(const struct device *dev, bt_hci_recv_t recv)
+	/* Abort RX thread */
+	k_thread_abort(cfg->rx_thread);
+
+	return 0;
```
done

lfilliot commented:
Hello @jhedberg , @cvinayak ,
Thanks for the approvals ! What are the next steps to get this PR merged ? Are we waiting for more reviews ?
I can see one test failed, but this does not seem related to the PR content and I do not have a clue on the fail cause,
Thank you,

Thalley commented:
> Hello @jhedberg , @cvinayak , Thanks for the approvals ! What are the next steps to get this PR merged ? Are we waiting for more reviews ? I can see one test failed, but this does not seem related to the PR content and I do not have a clue on the fail cause, Thank you,

I've re-run the failing tests, as they should be passing in CI now. 
The next step is to wait until the current feature freeze is lifted and the 4.1 has been released (planned for Friday this week). After that, unless someone rejects this change, you should see this being merged shortly after (probably next week assuming that there's no git conflicts)
